core/src/Cuda/Kokkos_Cuda_Parallel.hpp

```diff
@@ -1992,14 +1994,22 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
     cudaFuncAttributes attr =
         CudaParallelLaunch<ParallelReduce,
                            LaunchBounds>::get_cuda_func_attributes();
-    m_team_size =
-        m_team_size >= 0
-            ? m_team_size
-            : Kokkos::Impl::cuda_get_opt_block_size<FunctorType, LaunchBounds>(
-                  m_policy.space().impl_internal_space_instance(), attr,
-                  m_functor, m_vector_size, m_policy.team_scratch_size(0),
-                  m_policy.thread_scratch_size(0)) /
-                  m_vector_size;
+
+    // Valid team size not provided, deduce team size
+    if (m_team_size < 0) {
+      m_team_size =
+          Kokkos::Impl::cuda_get_opt_block_size<FunctorType, LaunchBounds>(
+              m_policy.space().impl_internal_space_instance(), attr, m_functor,
+              m_vector_size, m_policy.team_scratch_size(0),
+              m_policy.thread_scratch_size(0)) /
+          m_vector_size;
+
+      // Currently we require power-of-two team size for reductions.
+      int pow2 = 0;
+      while (m_team_size >>= 1 > 0) pow2++;
+      pow2        = 1 << pow2;
+      m_team_size = pow2;
```

Comment:
I don't think this is the right approach. I think the cuda_get_opt_block_size function needs an option to only return power of two (if that actually is the issue). Because you might defeat the purposes of opt block size with what you are doing here.

Reply:
Is m_vector_size always a pow2 number?

Reply:
Yes. I'll add that. That is the issue.